tensorboard/plugins/profile/profile_plugin.py

```diff
@@ -0,0 +1,167 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""The TensorBoard plugin for performance profiling."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import logging
+import os
+import tensorflow as tf
+from werkzeug import wrappers
+
+from tensorboard.backend import http_util
+from tensorboard.backend.event_processing import plugin_asset_util
+from tensorboard.plugins import base_plugin
+from tensorboard.plugins.profile import trace_events_json
+from tensorboard.plugins.profile import trace_events_pb2
+
+# The prefix of routes provided by this plugin.
+_PLUGIN_PREFIX_ROUTE = 'profile'
+
+# HTTP routes
+LOGDIR_ROUTE = '/logdir'
+DATA_ROUTE = '/data'
+RUNS_ROUTE = '/tags'
+
+# Available profiling tools -> file name of the tool data.
+_FILE_NAME = 'TOOL_FILE_NAME'
+TOOLS = {
+    'trace_viewer': 'trace',
+}
+
+
+def process_raw_trace(raw_trace):
+  """Processes raw trace data and returns the UI data."""
+  trace = trace_events_pb2.Trace()
+  trace.ParseFromString(raw_trace)
+  return ''.join(trace_events_json.TraceEventsJsonStream(trace))
+
+
+class ProfilePlugin(base_plugin.TBPlugin):
+  """Profile Plugin for TensorBoard."""
+
+  plugin_name = _PLUGIN_PREFIX_ROUTE
+
+  def __init__(self, context):
+    """Constructs a profiler plugin for TensorBoard.
+
+    This plugin adds handlers for performance-related frontends.
+
+    Args:
+      context: A base_plugin.TBContext instance.
+    """
+    self.logdir = context.logdir
+    self.plugin_logdir = plugin_asset_util.PluginDirectory(
+        self.logdir, ProfilePlugin.plugin_name)
+
+  @wrappers.Request.application
+  def logdir_route(self, request):
+    return http_util.Respond(request, {'logdir': self.plugin_logdir},
+                             'application/json')
+
+  def _run_dir(self, run):
+    run_dir = os.path.join(self.plugin_logdir, run)
+    return run_dir if tf.gfile.IsDirectory(run_dir) else None
+
+  def runs_impl(self):
+    """Returns available runs and available tool data in the log directory.
+
+    In the plugin log directory, each directory contains profile data for a
+    single run (identified by the directory name), and files in the run
+    directory contains data for different tools. The file that contains profile
+    for a specific tool "x" will have a fixed name TOOLS["x"].
+    Example:
+      log/
+        run1/
+          trace
+        run2/
+          trace
+
+    Returns:
+      A map from runs to tool names e.g.
+        {"run1": ["trace_viewer"], "run2": ["trace_viewer"]} for the example.
+    """
+    runs = {}
+    if not tf.gfile.IsDirectory(self.plugin_logdir):
+      return runs
+    for run in tf.gfile.ListDirectory(self.plugin_logdir):
```

Comment:
This isn't how the plugin directory was intended to be used. It looks like you're expecting all plugin data for every run to be located in a single plugins folder, which is separate from the runs. As in the following example:

```
root_logdir/
  plugins/
    profile/
      run1/
      run2/
  run1/
    ...
  run2/
   ....
```

However, this is not robust; if, for example, you point tensorboard just at an individual run (logdir=root_logdir/run1) then we lose access to the plugin data. Instead, the following directory structure is intended:

```
root_logdir/
  run1/
    plugins/
      profile/
      trace
  run2/
    plugins/
      profile/
        trace
```
This way, all run data is always packaged within that run's folder.

For implementation, you would need to iterate over every subdirectory of the root directory, and test whether it has the plugin_logdir you're interested in. 

Fortunately, the EventMultiplexer is able to do this all for you. See: [PluginAssets](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/event_processing/event_multiplexer.py#L202) on the multiplexer. It handles the run multiplexing itself.